### ISSUE TITLE
StackOverflowError when using parSequence

### ISSUE THREAD

**Issue by oleg-py** [monix/monix#1033]
Running this app:

```scala
import monix.eval.{Task, TaskApp}
import cats.effect.ExitCode
import cats.implicits._

object TestRunner extends TaskApp {
  override def run(args: List[String]): Task[ExitCode] = {
    List.fill(2000)(Task.unit).parSequence_
      .executeWithOptions(_.enableLocalContextPropagation)
      .as(ExitCode.Success)
  }
}
```

produces (shortened):

```
java.lang.StackOverflowError
        at java.security.AccessController.doPrivileged(Native Method)
        at java.lang.ClassLoader.checkPackageAccess(ClassLoader.java:500)
        at monix.execution.internal.Trampoline.monix$execution$internal$Trampoline$$immediateLoop(Trampoline.scala:70)

        // :loop
        at monix.execution.internal.Trampoline.startLoop(Trampoline.scala:32)
        at monix.execution.schedulers.TrampolineExecutionContext$JVMOptimalTrampoline.startLoop(TrampolineExecutionContext.scala:143)
        at monix.execution.internal.Trampoline.execute(Trampoline.scala:40)
        at monix.execution.schedulers.TrampolineExecutionContext.execute(TrampolineExecutionContext.scala:64)
        at monix.execution.schedulers.BatchingScheduler.execute(BatchingScheduler.scala:50)
        at monix.execution.schedulers.BatchingScheduler.execute$(BatchingScheduler.scala:47)
        at monix.execution.schedulers.TracingScheduler$Base.execute(TracingScheduler.scala:53)
        at monix.eval.internal.TaskRestartCallback.start(TaskRestartCallback.scala:56)
        at monix.eval.internal.TaskRunLoop$.executeAsyncTask(TaskRunLoop.scala:592)
        at monix.eval.internal.TaskRunLoop$.startFull(TaskRunLoop.scala:120)
        at monix.eval.Task$.unsafeStartNow(Task.scala:4582)
        at monix.eval.Task$.unsafeStartEnsureAsync(Task.scala:4563)
        at monix.eval.internal.TaskMapBoth$Register.apply(TaskMapBoth.scala:128)
        at monix.eval.internal.TaskMapBoth$Register.apply(TaskMapBoth.scala:43)
        at monix.eval.internal.TaskRestartCallback.run(TaskRestartCallback.scala:65)
        at monix.execution.internal.Trampoline.monix$execution$internal$Trampoline$$immediateLoop(Trampoline.scala:66)
        // goto loop
```

**Comment by Avasil** [comment]
Does it happen without `executeWithOptions`? 

I think `parSequence` is quite inefficient (IIRC it just folds everything with `parMap2` and it comes from cats type classes so we can't override it) and I assume extra garbage created from constant wrapping of `Scheduler` could be too much.

**Comment by alexandru** [comment]
Regardless of how efficient `parSequence` is, it should be stack safe. 

We'll have to check that loop, I might have missed introducing a trampolined boundary somewhere.

**Comment by oleg-py** [comment]
It happens only when `executeWithOptions` sets `enableLocalContextPropagation`. If you do `executeWithOptions(x => x)`, it won't crash.

The garbage isn't the problem, I wasn't looking for memory issues :)

**Comment by alexandru** [comment]
> If you do executeWithOptions(x => x), it won't crash.

I believe we have an internal optimization for that 😄

**Comment by Avasil** [comment]
I didn't have a chance to look at it closer but if we use `Scheduler.traced` then StackOverflow is gone. Might be connected to #1023 

I wonder if it would be possible to reuse `TracingScheduler` reference instead of creating it each `TaskShift` / `Context` . Although it wouldn't fix the root of the problem, only symptoms

**Comment by Avasil** [comment]
Yeah, I'm pretty sure it's connected. I don't know why exactly it stack overflows but it has a memory leak for sure. 

I don't have any experience in debugging such issues so I'm not making a lot of progress but here's a simple reproduction example:

```scala
import cats.effect.ExitCode
import cats.implicits._
import monix.eval.{Task, TaskApp}

object TestRunner extends TaskApp {

  override def run(args: List[String]): Task[ExitCode] = {
    parSequence(List.fill(200)(Task.unit))
      .loopForever
      .executeWithOptions(_.enableLocalContextPropagation)
      .as(ExitCode.Success)
  }

  def parSequence(a: List[Task[Unit]]): Task[Unit] =
    a.foldLeft[Task[Any]](Task.unit)((x, y) => Task.parZip2(x, y)).void
}
```

with `javaOptions += "-Xmx64m"` it will throw `OOM`

Heap Dump:

https://drive.google.com/open?id=1xKRcC2HlQkX0aohPWqqeCkt05nvlht-l

**Comment by Avasil** [comment]
Should be fixed in `3.1.0`
